Logging Pt.2: Predefined categories and usage examples

uklotzde opened the pull request:
***TODO***
- [ ] Merge #3073 into *master* to hide all unrelated changes from the diff

The documentation can be found in logging.h.

- Add predefined root and top-level logging categories
- Provide a macro for creating custom child categories
- Provide 2 reference examples (now using Info instead of Debug logs):
    - SoundSourceFFmpeg
    - SoundDevicePortAudio
- Add missing thread names

SoundDevicePortAudio logs:
```
Debug [Main]: SoundManager::setupDevices()
Info [Main] mixxx.device.PortAudio: Opening device "default"
Info [Main] mixxx.device.PortAudio: Number of output channels: 2
Info [Main] mixxx.device.PortAudio: Number of input channels: 0
Info [Main] mixxx.device.PortAudio: Requested sample rate: 44100 Hz
Info [Main] mixxx.device.PortAudio: Number of sample frames per buffer: 1024
Info [Main] mixxx.device.PortAudio: Latency per buffer: 23.22 ms
Info [Main] mixxx.device.PortAudio: Opening stream 14
Debug [Controller]: Found "" "" "r1" S/N "" "Interface 0"
Warning [Controller]: USB permissions problem (or device error.) Your account needs write access to USB HID controllers.
[New Thread 0x7ffeb38d1700 (LWP 14515)]
Debug [Controller]: Found "" "" "r1" S/N "" "Interface 1"
Warning [Controller]: USB permissions problem (or device error.) Your account needs write access to USB HID controllers.
Debug [Controller]: Found "" "" "r4624" S/N "" "Interface 0"
Warning [Controller]: USB permissions problem (or device error.) Your account needs write access to USB HID controllers.
Debug [Controller]: Found "" "" "r4624" S/N "" "Interface 1"
Warning [Controller]: USB permissions problem (or device error.) Your account needs write access to USB HID controllers.
Debug [Controller]: Found "" "" "r4624" S/N "" "Interface 2"
Warning [Controller]: USB permissions problem (or device error.) Your account needs write access to USB HID controllers.
Debug [Controller]: ControllerManager::getControllerList
Debug [Controller]: Controller polling stopped.
Debug [Main] mixxx.device.PortAudio: "SoundDeviceId(default, 14)" Opened stream
Info [Main] mixxx.device.PortAudio: Loading dynamic library "libportaudio.so.2"
Debug [Main] mixxx.device.PortAudio: Loaded dynamic library "libportaudio.so.2"
Info [Main] mixxx.device.PortAudio: Enabling real-time scheduling
[New Thread 0x7ffea70d0700 (LWP 14516)]
Debug [Main] mixxx.device.PortAudio: "SoundDeviceId(default, 14)" Stream started
Info [Main] mixxx.device.PortAudio: Actual sample rate: 44100 Hz
Info [Main] mixxx.device.PortAudio: Actual output latency: 46.4399 ms
Info [0x7fff00007720] mixxx.device.PortAudio: SSE: Enabling denormals to zero mode
Info [0x7fff00007720] mixxx.device.PortAudio: SSE: Enabling flush to zero mode
Info [0x7fff00007720] mixxx.device.PortAudio: Denormals to zero mode is working
```

SoundSourceFFmpeg logs:
```
Debug [AnalyzerThread 0 #1] mixxx.source.FFmpeg: Opened stream for decoding { index 0 | id 0 | codec_type 1 | codec_id 86017 | channels 2 | channel_layout 3 | channel_layout (fixed) 3 | format 8 | sample_rate 44100 | bit_rate 320000 | frame_size 1152 | initial_padding 0 | trailing_padding 0 | seek_preroll 0 | start_time 353600 | duration 2794291200 | nb_frames 0 | time_base 1 / 14112000 }
Debug [AnalyzerThread 0 #1] mixxx.source.FFmpeg: Lead-in frames [0 -> 1105)
[mp3float @ 0x7ffee0066f40] Could not update timestamps for skipped samples.
Debug [AnalyzerThread 0 #1]: AnalysisDAO fetched 0 analyses, 0 bytes for track 306621 in 0 ms
Debug [AnalyzerThread 0 #1]: Skipping AnalyzerEbur128
Debug [AnalyzerThread 0 #1]: AnalyzerBeats preference settings: 
```

uklotzde committed:
Predefine base Mixxx logging categories

uklotzde committed:
Use category logging for SoundSourceFFmpeg

uklotzde committed:
Use category logging for SoundDevicePortAudio

uklotzde committed:
Initialize thread names for better logging messages

uklotzde committed:
Add logging example references in documentation

Be-ing commented on src/soundio/sounddeviceportaudio.h:
```diff
@@ -26,17 +26,17 @@
-/** Dynamically resolved function which allows us to enable a realtime-priority callback
-    thread from ALSA/PortAudio. This must be dynamically resolved because PortAudio can't
-    tell us if ALSA is compiled into it or not. */
-typedef int (*EnableAlsaRT)(PaStream* s, int enable);
```
Is this unused? Why did you change it in this logging PR?

uklotzde replied:
Moved to the .cpp file, not needed in the header.

Holzhaus commented on src/util/logging.cpp:
```diff
@@ -17,6 +16,18 @@
 
+Q_LOGGING_CATEGORY(mixxxLog, MIXXX_LOGGING_CATEGORY_ROOT)
+
+Q_LOGGING_CATEGORY(mixxxLogConfig, MIXXX_LOGGING_CATEGORY_CONFIG)
```
Is there any advantage from defining them all in one place?
Personally, I think all these macros look ugly.

I'd prefer this:

> ```cpp
> bool Highscore::load(const QString &filepath)
> {
>     QLoggingCategory fcIo("fc.io");
>     qCDebug(fcIo) << "Loading highscore ...";
>     QFile file(filepath);
>     if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
>         qCWarning(fcIo) << "Could not open file " << filepath << "!";
>         return;
>     }
>     // ...
> }
>```
>
> [...]
>
> QLoggingCategory represents a category in code, in this case it's "fc.io". It's not a one to one mapping though ... you can have multiple instances representing the same category.

Source: https://www.qt.io/blog/2014/03/11/qt-weekly-1-categorized-logging

uklotzde replied:
The idea was to group the logs under some predefined top-level categories to allow basic filtering without needing to know all child categories. The custom macro MIXXX_LOGGING_CATEGORY_PARENT_CHILD establishes a naming convention.

The macros Q_DECLARE_LOGGING_CATEGORY/Q_LOGGING_CATEGORY help to find all categories that are available, even if they are only defined locally inside a .cpp file and not intended to be reused.

Holzhaus replied:
> The idea was to group the logs under some predefined top-level categories to allow basic filtering without needing to know all child categories. The custom macro MIXXX_LOGGING_CATEGORY_PARENT_CHILD establishes a naming convention.

I thought this is done implicitly by separating the category name with dots.
And filtering can be done like this:

    QT_LOGGING_RULES="mixxx.engine.controls.*=true"

> The macros Q_DECLARE_LOGGING_CATEGORY/Q_LOGGING_CATEGORY help to find all categories that are available, even if they are only defined locally inside a .cpp file and not intended to be reused.

So you can list all categories via command line? Or do you mean by grepping through the source?

Holzhaus replied:
@uklotzde can you elaborate?

Holzhaus replied:
Ping.

uklotzde replied:
Without the constants all the logging categories would be scattered in the code as arbitrary, redundant strings. Renaming or restructuring them would be really tedious.

The Q_DECLARE_LOGGING_CATEGORY macro explicitly applies the *separated by dots* convention. Otherwise there is no rule how to structure the names of logging categories.

Holzhaus replied:
OK, sounds good. I'll give this a test. Hopefully we can stop sprinkling commented out qdebug calls all over our code :smile:

uklotzde replied:
I don't like to introduce a custom macro. But so far I didn't have any idea how to achieve this in a better way.

Holzhaus replied:
We could put all category strings in a single header file and then include them from there. That way you only need to seach-replace in a single file.

uklotzde replied:
With this restriction the macros could be avoided.

uklotzde replied:
After 2nd thought: Not viable. A huge global file that might contain a separate logging category for each module (comprising 1 or a few .cpp files) is unmaintainable.

How is a user supposed to chose the correct one? Even the list of predefined top-level logging categories in logging.h is long already.

Holzhaus replied:
Hmm. I also thought about this and I'm not even sure we need this global list. When we define the logging category inside the source file, like this it becomes much easier to find the correct logging category to enable debug messages for:

    QLoggingCategory logcat("device.soundsource.portaudio");

You're right that it becomes a bit tedious to restructure logging categories later on if the fully qualified names are spread across the code base like this, but it's still possible to mass-replace them in a bash one-liner using `find`/`grep`/`sed`. And I don' t think we want to restructure them that often, and the instant visibility of the full logging category name when reading the source file is very convenient and outweighs this potential issue IMHO.

uklotzde committed:
Merge branch '2.3' of git@github.com:mixxxdj/mixxx.git

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
Merge branch 'master' into logging

daschuer commented:
The version bump does not help. 

This fails now. 
subprocess.CalledProcessError: Command '['clang-format', '--dump-config', '/home/travis/build/mixxxdj/mixxx/tools/../src/mixxx.cpp']' returned non-zero exit status 1.

It leads that to the line-length check reformat the files with clan-format default styling. 
This results into a toggling complains either of the line-lenght or the clang-format hook.

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
Merge branch 'master' into logging

Holzhaus commented on src/sources/soundsourceffmpeg.cpp:
```diff
@@ -79,15 +79,15 @@ inline int64_t getStreamStartTime(const AVStream& avStream) {
         // This case is not unlikely, e.g. happens when decoding WAV files.
-#if ENABLE_TRACING
-        kLogger.trace()
+#if ENABLE_TRACE_LOG
```
Can't we get rid of this?

I don't think this makes a notable performance improvement:
```
Note: Arguments aren't processed if the debug output for that category is not enabled, so don't rely on any side effects.
```

uklotzde replied:
Qt does not provided a detailed trace level, only debug. Until the default log level has been switched from *debug* to *info* we cannot avoid this conditional compilation to avoid log spam.

And even then log messages in real-time threads need to be disabled at compile time by using conditional compilation. That's the reason why I also picked an example from the sound device code.

Holzhaus replied:
Why do we need something more detailed than debug? IMHO it makes more sense to switch it on for the logging category you're interested in and use INFO for everything else by default.

Why do they need to be disabled in RT code? If arguments aren't processed if debug logging is disabled (as the documentation states), shouldn't this just cause one additional conditional branch per log statement? That should be OK even in performance-critical code, right?

Be-ing replied:
String operations are too slow for realtime code.

uklotzde replied:
The logs we are talking about in the real-time code are only used for tracing execution paths in the callback loop during development. These are invoked frequently and should **not** be compiled into any executable version, even if they would be real-time safe.

The Qt logging facilities will have some performance impact at runtime. And they have been designed for general purpose desktop UI applications, not for performance critical code.

uklotzde replied:
I will add a global define to enable verbose debug logs like in `SoundSourceFFmpeg`. These could be enabled permanently once we have switched the default log level from *Debug* to *Info*.

The local defines for trace logs in `SoundDevicePortAudio` are still needed as explained. I will try to add a reference to this example as blueprint.

uklotzde committed:
Introduce global define for verbose debug logging

uklotzde committed:
Add warning about trace logging in real-time code

uklotzde committed:
Merge branch 'master' of git@github.com:mixxxdj/mixxx.git into logging

Be-ing commented on src/util/logging.h:
```diff
@@ -1,9 +1,91 @@
+Q_DECLARE_LOGGING_CATEGORY(mixxxLogEngine)
+#define MIXXX_LOGGING_CATEGORY_ENGINE \
+    MIXXX_LOGGING_CATEGORY_PARENT_CHILD(MIXXX_LOGGING_CATEGORY_ROOT, "engine")
+
```
```suggestion
/// Base logging category for effects
Q_DECLARE_LOGGING_CATEGORY(mixxxLogEffects)
#define MIXXX_LOGGING_CATEGORY_ENGINE \
    MIXXX_LOGGING_CATEGORY_PARENT_CHILD(MIXXX_LOGGING_CATEGORY_ROOT, "effects")
```

I can then use this in #2618 after this is merged.

Be-ing commented on src/util/logging.h:
```diff
@@ -1,9 +1,91 @@
+#define MIXXX_LOGGING_CATEGORY_NETWORK \
+    MIXXX_LOGGING_CATEGORY_PARENT_CHILD(MIXXX_LOGGING_CATEGORY_ROOT, "network")
+
+/// Base logging category for scripting
+Q_DECLARE_LOGGING_CATEGORY(mixxxLogScript)
+#define MIXXX_LOGGING_CATEGORY_SCRIPT \
+    MIXXX_LOGGING_CATEGORY_PARENT_CHILD(MIXXX_LOGGING_CATEGORY_ROOT, "script")
```
```suggestion
/// Base logging category for controllers
Q_DECLARE_LOGGING_CATEGORY(mixxxLogController)
#define MIXXX_LOGGING_CATEGORY_SCRIPT \
    MIXXX_LOGGING_CATEGORY_PARENT_CHILD(MIXXX_LOGGING_CATEGORY_ROOT, "controller")
```

There is info logged by the controller system apart from scripts. ControllerEngine could create a mixxx.controller.CONTROLLERNAME.script subcategory for messages from the script.

Be-ing commented on src/util/logging.h:
```diff
@@ -20,8 +102,32 @@ constexpr LogLevel kLogLevelDefault = LogLevel::Warning;
+/// If you need more detailed trace logs during debugging or implementing
+/// new functionality then hide this code behind `#define ENABLE_TRACE_LOG false`
+/// in the corresponding compilation unit (= .cpp file). Enable it temporarily
+/// in your local repo while debugging. Disable it before submitting a pull
+/// request. Use qCDebug() for these trace logs.
```
```suggestion
/// in your local repo while debugging. Disable it before committing.
/// Use qCDebug() for these trace logs.
```

Be-ing commented on src/util/logging.h:
```diff
@@ -1,9 +1,91 @@
+#define MIXXX_LOGGING_CATEGORY_DATABASE \
+    MIXXX_LOGGING_CATEGORY_PARENT_CHILD(MIXXX_LOGGING_CATEGORY_ROOT, "database")
+
+/// Base logging category for hardware I/O sound devices
+Q_DECLARE_LOGGING_CATEGORY(mixxxLogDevice)
+#define MIXXX_LOGGING_CATEGORY_DEVICE \
+    MIXXX_LOGGING_CATEGORY_PARENT_CHILD(MIXXX_LOGGING_CATEGORY_ROOT, "device")
```
```suggestion
/// Base logging category for hardware I/O sound devices
Q_DECLARE_LOGGING_CATEGORY(mixxxLogAudioDevice)
#define MIXXX_LOGGING_CATEGORY_DEVICE \
    MIXXX_LOGGING_CATEGORY_PARENT_CHILD(MIXXX_LOGGING_CATEGORY_ROOT, "audiodevice")
```

"Device" by itself could mean lots of things.

uklotzde committed:
Merge branch 'master' into logging

# Conflicts:
#	src/sources/soundsourceffmpeg.cpp

uklotzde committed:
SoundSourceFFmpeg: Rename variables and reformat code

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
Merge branch 'master' into logging

# Conflicts:
#	src/sources/soundsourceffmpeg.cpp

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
SoundSourceFFmpeg: Reorder and document decoding loop body

- Document and illustrate all steps in detail
- Verify assumptions by debug assertions

uklotzde committed:
Replace log level Trace with Debug

uklotzde committed:
Fix debug assertion when decoding corrupt file

uklotzde committed:
Merge branch 'master' into logging

uklotzde committed:
Merge branch '2.3' into logging

uklotzde committed:
Merge branch '2.3_ffmpeg' into logging

# Conflicts:
#	src/sources/soundsourceffmpeg.cpp

ywwg commented:
Can we have an update on this PR? Do people like where it is going? What do we need to do to make it not a draft?

uklotzde commented:
I am not able to continue working on this until other PRs have been finished and merged.

uklotzde committed:
Merge branch 'main' into logging

# Conflicts:
#	src/soundio/sounddeviceportaudio.cpp
#	src/soundio/sounddeviceportaudio.h
#	src/sources/soundsourceffmpeg.cpp
#	src/sources/soundsourceffmpeg.h

uklotzde commented:
I am tired to work on C++ code and don't plan to revive this PR.

uklotzde commented:
I will delete the corresponding branches in my fork eventually to get rid of the cognitive load.
